### PR TITLE
feat: replace Makefile, add yapf/isort to pre-commit, other small improvements

### PR DESCRIPTION
**Pull request by hckr**

### CODE REVIEW

File: {{ cookiecutter.repo_name }}/.flake8 [hunk 1]
```diff
@@ -1,3 +1,3 @@
 [flake8]
 max-line-length = 120
-ignore = F841,E722,W605,E265
+ignore = F841,E722,W605,E265,E501
```
Comment (jaroslawk) [inline {{ cookiecutter.repo_name }}/.flake8:3]: do we actually need flake? most of the time it signals the same information than `pylint`

File: {{ cookiecutter.repo_name }}/.github/workflows/lint.yml [hunk 2]
```diff
@@ -9,10 +9,9 @@ on:
 
 env:
   python-version: "3.10"
-  docker-registry: 003209562630.dkr.ecr.eu-central-1.amazonaws.com
+  docker-registry: XXX.dkr.ecr.eu-central-1.amazonaws.com
```
Comment (jaroslawk) [inline {{ cookiecutter.repo_name }}/.github/workflows/lint.yml]: actually, this should be always `003209562630` ... guess this is our ECR address
@mattarkowski can this be anything else?

File: {{ cookiecutter.repo_name }}/.github/workflows/lint.yml [hunk 3]
```diff
@@ -44,7 +43,7 @@ jobs:
           path: ~/.cache/pypoetry/virtualenvs
           key: ${{ runner.os }}-python-${{ steps.setup-python.outputs.python-version }}-poetry-${{ hashFiles('**/poetry.lock', '**/.pre-commit-config.yaml') }}
```
Comment (jaroslawk) [inline {{ cookiecutter.repo_name }}/.github/workflows/lint.yml]: let's change those also - there is already support for poetry setup-python:

https://github.com/NaturalAntibody/abstudio-cli/blob/cecd3d2316f68a6824b4281958deb9616c5f9d21/.github/workflows/cicd.yml#L35

we can remove our support for poetry cache

File: {{ cookiecutter.repo_name }}/.pre-commit-config.yaml [hunk 4]
```diff
@@ -16,15 +16,28 @@ repos:
...
+        args: [--py39-plus]
+
   - repo: local
     hooks:
+      - id: isort
+        name: isort
+        entry: isort
+        language: python
+        types: [python]
+      - id: yapf
```
Comment (jaroslawk) [inline {{ cookiecutter.repo_name }}/.pre-commit-config.yaml:35]: 💯

File: {{ cookiecutter.repo_name }}/README.md [hunk 5]
```diff
@@ -1,10 +1,33 @@
...
 
 {{cookiecutter.description}}
 
-Project Organization
-------------
+## How to start
+
+Init the git repo (e.g. with `git init`).
+
+`./create_environment` or if you prefer, run these
```
Comment (jaroslawk) [inline {{ cookiecutter.repo_name }}/README.md]: `./create_environment => ./create_environment.sh`

File: {{ cookiecutter.repo_name }}/cli.py [hunk 6]
```diff
@@ -0,0 +1,33 @@
+import subprocess
```
Comment (jaroslawk) [inline {{ cookiecutter.repo_name }}/cli.py]: * can we move `*.sh` to `bin`? 
* `cli.py => data_sync.py` ? anything less general than `cli`

File: {{ cookiecutter.repo_name }}/create_environment.sh [hunk 7]
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
+poetry install
+poetry run pre-commit install
```
Comment (jaroslawk) [inline {{ cookiecutter.repo_name }}/create_environment.sh]: no strong opinion here but since this is one-off action to setup a repo - maybe we should not focus on automating this action?

File: {{ cookiecutter.repo_name }}/{{ cookiecutter.repo_name }}/settings.py [hunk 8]
```diff
@@ -18,8 +18,6 @@
 
 load_dotenv()
 
-BASE = os.path.dirname(__file__)
-
 class Settings(TypedDict):
     DATA_PATH: Path
```
Comment (jaroslawk) [inline {{ cookiecutter.repo_name }}/{{ cookiecutter.repo_name }}/settings.py]: I am not fan of TypeDict can we use dataclass-es?

**Review by jaroslawk — COMMENTED** [review]
let's consider removal of flake8

**Comment by jaroslawk** [comment]
@tadsatlawa how about removing `flake8`?

File: {{ cookiecutter.repo_name }}/.github/workflows/lint.yml (hunk 2 above)
Comment (mattarkowski) [reply {{ cookiecutter.repo_name }}/.github/workflows/lint.yml]: imo whole repo url should be passed as a env or sth instead of plain text in repo

Comment (mattarkowski) [reply {{ cookiecutter.repo_name }}/.github/workflows/lint.yml]: this is how we do it in our backends 
```
      - name: Init terraform
        env:
          AWS_ACCESS_KEY_ID: ${{ secrets.AWS_ACCESS_KEY_ID }}
          AWS_SECRET_ACCESS_KEY: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
          AWS_REGION: ${{ env.aws-region }}
        working-directory: ./terraform
        run: |
          terraform init -backend-config="access_key=${{ secrets.AWS_ACCESS_KEY_ID }}" -backend-config="secret_key=${{ secrets.AWS_SECRET_ACCESS_KEY }}"
          terraform workspace select ${{ env.env }}
      - name: Get repository and store it in env var
        working-directory: ./terraform
        run: |
          REPOSITORY_URL=$(terraform output --json ecr_repository | jq -r '.repository_url')
          echo "REPOSITORY URL: $REPOSITORY_URL"
          echo "repository_url=$REPOSITORY_URL" >> $GITHUB_ENV
```

File: {{ cookiecutter.repo_name }}/{{ cookiecutter.repo_name }}/settings.py (hunk 8 above)
Comment (hckr) [reply {{ cookiecutter.repo_name }}/{{ cookiecutter.repo_name }}/settings.py]: It was proposed by @pdeszynski

File: {{ cookiecutter.repo_name }}/create_environment.sh (hunk 7 above)
Comment (hckr) [reply {{ cookiecutter.repo_name }}/create_environment.sh]: It was previosuly done in Makefile, I didn't want to loose this functionality. IMO this way it's less probable that someone forgets to install git hooks

File: {{ cookiecutter.repo_name }}/.github/workflows/lint.yml (hunk 2 above)
Comment (hckr) [reply {{ cookiecutter.repo_name }}/.github/workflows/lint.yml]: This repo doesn't mention terraform anywhere, so I'll just remove `docker-registry` altogether.

File: {{ cookiecutter.repo_name }}/README.md [hunk 9]
```diff
@@ -15,16 +38,12 @@ Project Organization
     │   ├── processed      <- The final, canonical data sets for modeling.
     │   └── raw            <- The original, immutable data dump.
     │
-    ├── docs               <- A default Sphinx project; see sphinx-doc.org for details
-    │
     ├── models             <- Trained and serialized models, model predictions, or model summaries
     │
     ├── notebooks          <- Jupyter notebooks. Naming convention is a number (for ordering),
     │                         the creator's initials, and a short `-` delimited description, e.g.
     │                         `1.0-jqp-initial-data-exploration`.
```
Comment (BartoszJanuszNA) [inline {{ cookiecutter.repo_name }}/README.md]: We are not using autor's initials anymore. Also add "scripts" folder under notebooks, for storing eg. python training scripts to be run in the background.

File: {{ cookiecutter.repo_name }}/create_environment.sh (hunk 7 above)
Comment (BartoszJanuszNA) [inline {{ cookiecutter.repo_name }}/create_environment.sh]: Missing `poetry run pre-commit install --hook-type commit-msg`

File: {{ cookiecutter.repo_name }}/{{ cookiecutter.repo_name }}/settings.py (hunk 8 above)
Comment (pdeszynski) [reply {{ cookiecutter.repo_name }}/{{ cookiecutter.repo_name }}/settings.py]: Both will have their advantage and disadvantage. For e.g. using dataclasses only it will be cumbersome to do updates (for e.g. changing only 1 setting from a dataclass during initialization.

There will be a more scaffolding with dataclasses, but at the end I was using both to have simple API for updates, and dataclass to store settings.



```python
BASE = Path(__file__).parent.parent


class UpdateSettings(TypedDict, total=False):
    DATA_PATH: Path

@dataclass
class Settings:
    DATA_PATH: Path
    OTHER_PATH: Path = field(init=False)

    def __post_init__(self):
        self.OTHER_PATH = Path(os.getenv("OTHER_PATH", self.PATH / "other_path"))

    def update(self, new_settings: UpdateSettings):
        for key, value in new_settings.items():
            setattr(self, key, value)

        self.__post_init__()


settings = Settings(
    DATA_PATH=BASE / "data_path",
)


def update_settings(new_settings: UpdateSettings):
    settings.update(new_settings)
```

File: {{ cookiecutter.repo_name }}/mypy.ini [hunk 10]
```diff
@@ -18,3 +20,36 @@ ignore_missing_imports = True
...
+ignore_missing_imports = True
+
+[mypy-natsort.*]
+ignore_missing_imports = True
+
+[mypy-tensorflow.*]
+ignore_missing_imports = True
+
+[mypy-tqdm.*]
+ignore_missing_imports = True
```
Comment (BartoszJanuszNA) [inline {{ cookiecutter.repo_name }}/mypy.ini:55]: Good, maybe add keras, transformers and pytorch as well.

**Comment by pdeszynski** [comment]
`flake8` adds additionally complexity analysis, what other tools does not provide. Code should be linted by pylint and flake8 as in all other repositories

File: {{ cookiecutter.repo_name }}/pyproject.toml [hunk 11]
```diff
@@ -1,52 +1,67 @@
...
 
+[tool.poetry.scripts]
+cli = 'cli:cli'
+
 [tool.poetry.dependencies]
 python = "^3.9"
+
 python-dotenv = ">=0.21.0"
 
 [tool.poetry.dev-dependencies]
```
Comment (BartoszJanuszNA) [inline {{ cookiecutter.repo_name }}/pyproject.toml:17]: have you updated all of those deps?

File: {{ cookiecutter.repo_name }}/pyproject.toml [hunk 12]
```diff
@@ -1,52 +1,67 @@
...
+description = "{{ cookiecutter.description }}"
 name = "{{ cookiecutter.repo_name }}"
 version = "{{ cookiecutter.version }}"
-description = "{{ cookiecutter.description }}"
-authors = ["{{ cookiecutter.author_name }}"]
 
 include = ["{{ cookiecutter.repo_name }}/py.typed"]
 
+[tool.poetry.scripts]
+cli = 'cli:cli'
```
Comment (pdeszynski) [inline {{ cookiecutter.repo_name }}/pyproject.toml]: use `{{ cookiecutter.repo_name }} = "{{ cookiecutter.repo_name }}.__main__:cli"` 

1. left side should be unique per repo or if you'll get clashes for commands
2. I suggest using __main__ because of that being a python standard: https://docs.python.org/3/library/__main__.html#main-py-in-python-packages

File: {{ cookiecutter.repo_name }}/.github/workflows/lint.yml (hunk 2 above)
Comment (pdeszynski) [reply {{ cookiecutter.repo_name }}/.github/workflows/lint.yml]: for me any will work, you can remove it if we do not expect by default building a docker file.
You can also hardcode this URL as this is AWS account ID and will  never change

**Comment by tadsatlawa** [comment]
> @tadsatlawa how about removing `flake8`?

As Piotr said - keep it

File: {{ cookiecutter.repo_name }}/{{ cookiecutter.repo_name }}/settings.py [hunk 13]
```diff
@@ -28,10 +26,10 @@ class UpdateSettings(Settings, total=False):
 
 def load_settings() -> Settings:
     return {
-        'DATA_PATH': Path(os.getenv("DATA_PATH", Path(BASE) / '..' / 'data')),
+        'DATA_PATH': Path(os.getenv('DATA_PATH', Path().absolute().parent / 'data')),
```
Comment (BartoszJanuszNA) [inline {{ cookiecutter.repo_name }}/{{ cookiecutter.repo_name }}/settings.py]: "When pathsegments is empty, the current directory is assumed" - this is not the same as directory of current file. I suggest: 
```suggestion
        'DATA_PATH': Path(os.getenv('DATA_PATH', Path(__file__).absolute().parent.parent / 'data')),
```

File: {{ cookiecutter.repo_name }}/create_environment.sh (hunk 7 above)
Comment (pdeszynski) [reply {{ cookiecutter.repo_name }}/create_environment.sh]: no need anymore if you have in your commit hooks file:

```
default_install_hook_types: [pre-commit, commit-msg]
```

File: {{ cookiecutter.repo_name }}/{{ cookiecutter.repo_name }}/settings.py (hunk 13 above)
Comment (pdeszynski) [reply {{ cookiecutter.repo_name }}/{{ cookiecutter.repo_name }}/settings.py]: yup, use file location - this will completely change when working directory is different. Until that's not a dependency - it will probaby work, but when you'll add this as a dependency to another project - it will not search in a package, but from where `python` was invoked

File: {{ cookiecutter.repo_name }}/pyproject.toml (hunk 11 above)
Comment (hckr) [reply {{ cookiecutter.repo_name }}/pyproject.toml:17]: As they are all dev deps, I think about setting all versions to `"*"`

Comment (pdeszynski) [reply {{ cookiecutter.repo_name }}/pyproject.toml:17]: Don't do it because of resolve time drastic increase when doing updates. Make them >= and update them now and then to a newer version. 

Python package managers tend to download all matching versions before doing dependency resolution

File: {{ cookiecutter.repo_name }}/mypy.ini (hunk 10 above)
Comment (hckr) [reply {{ cookiecutter.repo_name }}/mypy.ini:55]: Done

File: pyproject.toml [hunk 14]
```diff
@@ -1,19 +1,21 @@
...
 
 [tool.poetry.dependencies]
 python = "^3.9"
+
+click = "^8.0.1"
+cookiecutter = "^1.7.3"
 mkdocs = "^1.2.3"
 mkdocs-cinder = "^1.2.0"
-cookiecutter = "^1.7.3"
 pytest = ">=7.2"
```
Comment (jaroslawk) [inline pyproject.toml:15]: dev dependencies...

**Review by jaroslawk — APPROVED** [review]
LGTM

File: {{ cookiecutter.repo_name }}/README.md [hunk 15]
```diff
@@ -29,46 +28,53 @@ Here's an instruction how to do it: https://code.visualstudio.com/docs/python/en
...
+├── data
+│   ├── external           <- Data from third party sources.
+│   ├── interim            <- Intermediate data that has been transformed.
+│   ├── processed          <- The final, canonical data sets for modeling.
+│   └── raw                <- The original, immutable data dump.
+│
+├── models                 <- Trained and serialized models, model predictions, or model summaries
+│
+├── notebooks              <- Jupyter notebooks. Naming convention is a number (for ordering),
+|                             and a short `-` delimited description, e.g.
```
Comment (BartoszJanuszNA) [inline {{ cookiecutter.repo_name }}/README.md:44]: Why we use `-` as word delimiter in notebook file names?

File: {{ cookiecutter.repo_name }}/README.md [hunk 16]
```diff
@@ -29,46 +28,53 @@ Here's an instruction how to do it: https://code.visualstudio.com/docs/python/en
...
+│                             `1.0-jqp-initial-data-exploration`.
```
Comment (BartoszJanuszNA) [inline {{ cookiecutter.repo_name }}/README.md]: `1.0-initial-data-exploration.ipynb`

File: {{ cookiecutter.repo_name }}/README.md (hunk 15 above)
Comment (hckr) [reply {{ cookiecutter.repo_name }}/README.md:44]: tbh I usually used names with underscores: like "1.0-generate_input_sequence" 😐 

I don't know why – it's taken from original cookiecutter-data-science
